### ISSUE TITLE
Windows.Forensics.SRUM artifact fails to parse database records

### ISSUE THREAD

**Issue by chris-counteractive** [Velocidex/velociraptor#690]
**summary:** Windows.Forensics.SRUM artifact fails to parse database records.

**environment:** 

- Microsoft Windows [Version 10.0.19041.572]
- velociraptor v0.5.0-1, latest release (still present as of latest commit at time of submission (f632706))

**steps to reproduce:**

1. run `velociraptor gui`
1. launch the `Windows.Forensics.SRUM` artifact
1. note the db itself is uploaded, though it appears incomplete, specifically missing `SruDbIdMapTable` records (see analysis below)
1. note there are no parsed results: 0 rows for execution stats, application resource usage, and network connections
1. errors show in the logs:

    ```text
    vql: Starting query execution.
    Time 0: Windows.Forensics.SRUM/Upload: Sending response part 0 215 B (1 rows).
    vql: Parsing SruDbIdMapTable for c:\windows\system32\sru\srudb.dat
    vql: Parsed c:\windows\system32\sru\srudb.dat successfully with 0 records
    vql: Symbol AppId not found. Current Scope is: [NULL, __destructors], [SRUMLocation, accessor, ExecutableRegex, NetworkConnectionsGUID, ApplicationResourceUsageGUID, ExecutionGUID, server_config, config, $cache, $acl, $uploader, Artifact, $root], [$throttle], [Windows_Forensics_SRUM_Upload_0_0], [Windows_Forensics_SRUM_Execution_Stats_1_0], []
    vql: srum_lookup_id: Field Id should be an int.
    Time 0: Windows.Forensics.SRUM/Execution Stats: Sending response part 0 0 B (0 rows).
    Time 0: Windows.Forensics.SRUM/Application Resource Usage: Sending response part 0 0 B (0 rows).
    Time 0: Windows.Forensics.SRUM/Network Connections: Sending response part 0 0 B (0 rows).
    Uploaded 1 files.
    ```

**analysis**

the proximate cause seems to be that velociraptor sees `SruDbIdMapTable` erroneously having 0 records, which causes the app id lookups to fail (via `srum_lookup_id`).  This lack of records shows in both the uploaded db file as well as in live parsing, which may help narrow down the issue (or point to some user error on my part).

On my test machine, for example, [ese2csv](https://github.com/MarkBaggett/ese-analyst) on the "live" `srudb.dat` shows `SruDbIdMapTable` has 28,170 records:

```
PS C:\Users\chris\Work\test> .\ese2csv.exe --list-tables C:\Windows\System32\sru\SRUDB.dat
Processing File C:\Windows\System32\sru\srudb.dat
TABLE LISTING FOR ESE FILE C:\Windows\System32\sru\srudb.dat

Table MSysObjects has 262 records
Table MSysObjectsShadow has 262 records
Table MSysObjids has 41 records
Table MSysLocales has 7 records
Table SruDbIdMapTable has 28170 records
Table SruDbCheckpointTable has 0 records
Table {DA73FB89-2BEA-4DDC-86B8-6E048C6DA477} has 7987 records
Table {5C8CF1C7-7257-4F13-B223-970EF5939312} has 45726 records
Table {7ACBBAA3-D029-4BE4-9A7A-0885927F1D8F} has 4430 records
Table {973F5D5C-1D90-4944-BE8E-24B94231A174} has 31120 records
Table {D10CA2FE-6FCF-4F6D-848E-B2E99266FA89} has 197600 records
Table {DD6636C4-8929-4683-974E-22C046A43763} has 4960 records
Table {FEE4E14F-02A9-4550-B5CE-5FA2DA202E37} has 2476 records
Table {B6D82AF1-F780-4E17-8077-6CB9AD8A6FC4} has 13 records
Table {D10CA2FE-6FCF-4F6D-848E-B2E99266FA86} has 648 records
Table {FEE4E14F-02A9-4550-B5CE-5FA2DA202E37}LT has 20 records

Done listing C:\Windows\System32\sru\srudb.dat
```

That same tool on the db file uploaded by velociraptor via the Windows.Forensics.SRUM artifact shows the following:

```
PS C:\Users\chris\Work\test> .\ese2csv.exe --list-tables .\srudb-F.BU8SJR08FQJ8U.dat
Processing File srudb-f.bu8sjr08fqj8u.dat
TABLE LISTING FOR ESE FILE srudb-f.bu8sjr08fqj8u.dat

Table MSysObjects has 262 records
Table MSysObjectsShadow has 262 records
Table MSysObjids has 41 records
Table MSysLocales has 7 records
Table SruDbIdMapTable has 0 records
Table SruDbCheckpointTable has 0 records
Table {DA73FB89-2BEA-4DDC-86B8-6E048C6DA477} has 0 records
Table {5C8CF1C7-7257-4F13-B223-970EF5939312} has 228 records
Table {7ACBBAA3-D029-4BE4-9A7A-0885927F1D8F} has 829 records
Table {973F5D5C-1D90-4944-BE8E-24B94231A174} has 0 records
Table {D10CA2FE-6FCF-4F6D-848E-B2E99266FA89} has 5864 records
Table {DD6636C4-8929-4683-974E-22C046A43763} has 0 records
Table {FEE4E14F-02A9-4550-B5CE-5FA2DA202E37} has 192 records
Table {B6D82AF1-F780-4E17-8077-6CB9AD8A6FC4} has 0 records
Table {D10CA2FE-6FCF-4F6D-848E-B2E99266FA86} has 645 records
Table {FEE4E14F-02A9-4550-B5CE-5FA2DA202E37}LT has 20 records

Done listing srudb-f.bu8sjr08fqj8u.dat
```

Using the `parse_ese` VQL function directly on the live db at `C:\Windows\System32\sru\srudb.dat` returns record counts consistent with the second output, without error (_i.e._, using `parse_ese` on `MSysObjectsShadow` produces the expected 262 rows, and ~228 for the 5C8C GUID vs. the actual 45,726).  I've not been able to dig enough to sort why this might be - it's odd that it doesn't fail, but gives meaningful-but-different data for various tables.

In any case, the `srum_lookup_id` function fails without records in `SruDbIdMapTable`, which means the `AppId` column isn't set, which leads to that symbol not found error, and chaos reigns.

I'll keep looking, but wanted to get this in just in case there's an easy fix I'm not seeing.  Thanks as always!

**Comment by chris-counteractive** [comment]
**update:** this is only a problem when using the `ntfs` accessor.  

if you switch the accessor to `file` it works like a charm, unfortunately the default is `ntfs`. i'm not savvy enough to diagnose why it fails, and why it fails in this particular way, but there it is.

**workaround:** use `file` and update the artifact to default to `file`, though someone could still switch it manually to the one that doesn't work.

i can submit a quick PR with the artifact default updated, if that'd help.  thanks!

**Comment by chris-counteractive** [comment]
As a follow-up, i found the workaround by downloading the `srudb.dat` file using both accessors via the VFS browser and checking them with ese2csv: they both downloaded as precisely the same size files, but with different hashes (naturally, i suppose, since one worked and one didn't).  

i'm sure this applies in other situations, but the current [docs](https://www.velocidex.com/docs/vql_reference/filesystem_accessors/) don't make it clear when (or even that) there'd be such relevant differences when choosing one accessor over the other - they just emphasize avoiding locked files via `ntfs` (and sometimes performance).  To a first approximation, path and byte size seems a reasonable proxy for "is there a big difference between these files," but in this case it was a big deal - not sure the best way to know when accessor choice should matter and when it shouldn't.

The blog entry about SRUM uses the `ntfs` accessor apparently without issue, my concern that this mismatch indicates a bug somewhere deeper.  Thought I'd note it, in case it's simply a matter of me learning the edge cases.  Thanks again!

**Comment by chris-counteractive** [comment]
Last update for now, sorry for the volume: even when using the `file` accessor, the "successful" parsing shows the expected results for execution stats and application resource usage, but not for network connections (`{DD6636C4-8929-4683-974E-22C046A43763}`) - that is, it uploads 0 records, consistent with the count in the "broken" `ntfs` accessor copy of the db:

```
Time 0: Windows.Forensics.SRUM/Network Connections: Sending response part 0 0 B (0 rows)
```

Apologies for the back and forth, but it's just a tricky bug to bracket in - the workaround of using the `file` accessor doesn't 100% solve the issue. Thanks for your patience.

**Comment by scudette** [comment]
Thanks for the thorough analysis!

So there are two moving parts here - extracting the file using raw NTFS parsing is one step and the second is parsing the ESE structures. The tricky thing is that is it hard to compare because the SRUM db is always changing so if we extracted it with one tool then another there is no guarantee they are the same anyway. Maybe if we did it really quickly there is a better chance they are the same?

From your analysis above it looks more like the ntfs extraction is somehow going wrong since you get more results with the file accessor than ntfs. What would be helpful is the output of istat.exe for the srumdb file.

I just tried this on my system and I found that the srumdb file is in fact sparse which might add to the confusion:
![image](https://user-images.githubusercontent.com/3856546/96942919-da8d6f80-1519-11eb-8833-0d9d37076eed.png)

You can tell because the file size is not exactly the same as the uploaded bytes. If you export the file via the prepare a zip method, Velociraptor will pack the uploaded file as is and also attach the index for it
![image](https://user-images.githubusercontent.com/3856546/96942980-0ad50e00-151a-11eb-8718-c7d2383bf435.png)

It does this because the file might be huge but really sparse (e.g. the USN log file). OTOH if you download the file directly from the upload tab then it will pad the sparse areas.
![image](https://user-images.githubusercontent.com/3856546/96943043-3b1cac80-151a-11eb-95e2-e85710c0c6b6.png)

It is quite safe to use the file accessor instead of the ntfs these days and we should probably do this for most artifacts - this is because the "file" access drops back to the "ntfs" accessor when the file is not locked (the ese file is locked internally but not from an OS perspective). Using the OS apis to copy the file is faster and safer but that said we still want to track down any problems with the ntfs parser if we can see them.

Could you please check the output of istat and then repeat with (the mft id can be seen in fls.exe or in Velociraptor with the mft parser):
```
velociraptor.exe query "SELECT * FROM parse_ntfs_ranges(device='\\\\.\\c:', inode='81812-128-3')"
```

![image](https://user-images.githubusercontent.com/3856546/96943262-e62d6600-151a-11eb-9032-1b0687ff8260.png)

**Comment by chris-counteractive** [comment]
Thank you for the detailed discussion, all very helpful!  To clarify a bit:

1. The discussion about byte size comparisons was based on collecting the file via VFS, under both `ntfs` and `file`, not the artifact-collected copy of the DB which I did notice was "de-sparse-ified" with the attendant .idx file.  Definitely something to be aware of, and my discussion about that was a bit ambiguous, but we did notice that and tried to correct for it before filing the issue.
1. Your point's well taken that the content of the SRUM db is dynamic, but we didn't expect it to be identical - it was the 0-record tables that were unusual.  We've tested the "stability" of that file by looping over it at random time intervals on a few test systems we have not (yet) seen a situation where record counts drop to zero (with one exception, the `SruDbCheckpointTable` is often empty).  We actually haven't seen significant count fluctuations within seconds or even minutes.  It's usually reasonably stable, but perhaps we're missing something.

As requested, the istat data for the MFT entry (obtained via fls: `0|c:/Windows/System32/sru/SRUDB.dat|59957-128-6|...`):

```
PS C:\Users\chris\Work\test> .\sleuthkit-4.10.0-win32\bin\istat.exe -r \\.\C: 59957-128-6
MFT Entry Header Values:
Entry: 59957        Sequence: 1
$LogFile Sequence Number: 29852811031
Allocated File
Links: 1

$STANDARD_INFORMATION Attribute Values:
Flags: Archive
Owner ID: 0
Security ID: 983  (S-1-5-19)
Last User Journal Update Sequence Number: 3920934032
Created:        2020-05-19 21:22:13.270880100 (Central Daylight Time)
File Modified:  2020-10-23 08:35:00.104807800 (Central Daylight Time)
MFT Modified:   2020-10-23 08:35:00.104807800 (Central Daylight Time)
Accessed:       2020-10-23 09:11:38.427673500 (Central Daylight Time)

$FILE_NAME Attribute Values:
Flags: Archive
Name: SRUDB.dat
Parent MFT Entry: 237895        Sequence: 1
Allocated Size: 6422528         Actual Size: 6422528
Created:        2020-05-19 21:22:13.270880100 (Central Daylight Time)
File Modified:  2020-06-01 22:59:36.401277200 (Central Daylight Time)
MFT Modified:   2020-06-01 22:59:36.401277200 (Central Daylight Time)
Accessed:       2020-06-01 22:59:36.401277200 (Central Daylight Time)

$ATTRIBUTE_LIST Attribute Values:
Type: 16-0      MFT Entry: 59957        VCN: 0
Type: 48-4      MFT Entry: 59957        VCN: 0
Type: 128-0     MFT Entry: 19153        VCN: 0
Type: 128-0     MFT Entry: 59934        VCN: 2212
Type: 128-0     MFT Entry: 329708       VCN: 3824
Type: 128-0     MFT Entry: 24121        VCN: 19712

Attributes:
Type: $STANDARD_INFORMATION (16-0)   Name: N/A   Resident   size: 72
Type: $ATTRIBUTE_LIST (32-5)   Name: N/A   Resident   size: 192
Type: $FILE_NAME (48-4)   Name: N/A   Resident   size: 84
Type: $DATA (128-6)   Name: N/A   Non-Resident   size: 118996992  init_size: 118980608
  Starting address: 13979100, length: 2212
  Starting address: 13981312, length: 1612
  Starting address: 13982924, length: 13600
  Starting address: 14501632, length: 2288
  Starting address: 14503920, length: 9340
```

... and the corresponding velociraptor output (the full entry with the $DATA attribute extension, used above with istat, gave no results, so I ran it again with just the entry number):

```powershell
PS C:\Users\chris\Work\test> .\velociraptor-v0.5.0-1-windows-amd64.exe query "SELECT * FROM parse_ntfs_ranges(device='\\\\.\\C:', inode='59957-128-6')"
[]

PS C:\Users\chris\Work\test> .\velociraptor-v0.5.0-1-windows-amd64.exe query "SELECT * FROM parse_ntfs_ranges(device='\\\\.\\C:', inode='59957')"
[
 {
  "Offset": 0,
  "Length": 9060352,
  "IsSparse": false
 },
 {
  "Offset": 9060352,
  "Length": 6602752,
  "IsSparse": true
 },
 {
  "Offset": 15663104,
  "Length": 65077248,
  "IsSparse": true
 },
 {
  "Offset": 80740352,
  "Length": 38256640,
  "IsSparse": true
 }
]
```

Thanks again, let me know if there's anything else I can provide.

**Comment by chris-counteractive** [comment]
**update using v0.5.1 release:**

1. the new `auto` accessor specification reliably produces more results, consistent with the workaround numbers above.  it's very odd, but I got one single run to give network results, while the rest of my many tests still give only execution and application resource numbers and no network results.  this, in spite of the fact that the network results table (`{DD6636C4-8929-4683-974E-22C046A43763}`) consistently shows thousands of records when queried with other tools.
1. switching it back to `ntfs` still shows no parsed results (_i.e._, it still repros the bug).  
1. incidentally, trying to change `auto` to `ntfs` in the GUI for testing revealed two small issues:
    1. if you try to edit a parameter that has a default, the onchange handler will re-fill the text box with the default as soon as it's empty (workaround is to append the new value before deleting the old one, so the box never gets empty).  I haven't spent much time with the new UI code, but root maybe somewhere around here: https://github.com/Velocidex/velociraptor/blob/0bd0a04db98c2c90a9c10c1aecd4c0dd43eb7de2/gui/velociraptor/src/components/flows/new-collection.js#L296
    1. There's a typo in the PaginationBuilder ("Specify Resorces" vs. "Reso**u**rces") here: https://github.com/Velocidex/velociraptor/blob/0bd0a04db98c2c90a9c10c1aecd4c0dd43eb7de2/gui/velociraptor/src/components/flows/new-collection.js#L35

Probably not worth separate issues for tracking, but I can create them if it's helpful.  Thanks again for the engagement.

**Comment by scudette** [comment]
Thanks for reporting the GUI bug - I will open an issue for it - definitely funny :-). 

So to recap there seem to be two issues - the first is the ntfs parsing which seems to have some problems extracting the file. The second is the ese parsing and comparison with other tools.  

It is very possible that there are gaps in the ese parser and we can nail it down by applying the same parser on the same file with the external tools. I believe when we upload the file using the auto accessor it will use the file accessor because the file is not generally locked for reading so I expect the file to be uploaded perfectly correctly (through the API).

So there are two separate issues.  If you can share the ese file (even privately) I can see what the difference in processing with ese2csv.exe. 

For the ntfs we need to get more low level info about the way the ntfs parser is parsing the file. The parser has a stand alone tool here https://github.com/Velocidex/go-ntfs

You can get the parse of each mft entry like this:
```
ntfs.exe stat \\.\c: 81812 --verbose
```

This shows all the runs in each VCN entry (so in your case we have 4 mft entries 19153, 59934, 329708 and 24121).

You can also look at how the ntfs parser reconstructs the file using the runs command:
![image](https://user-images.githubusercontent.com/3856546/97067330-a42a2000-15ff-11eb-8ecf-f99809296096.png)

So for example in your case I would expect to have 4 top level readers (one for each VCN) then each will be broken into smaller readers for each run. This is how the ntfs parser works - it builds a reader tree which maps each block in the file to a reader responsible for it (which might in turn map to a run or a sparse null reader).

So it would be really nice to see the output of runs command which is way more detailed than the velociraptor output, and then maybe the output of the stat command on the 5 entries (the VCN ones and the original one).

Thanks again!

**Comment by scudette** [comment]
I just wanted to update this bug with my investigation of the ese parser. To double check the ese parser I collected the SRUM artifact and also saw that the network was returning 0 rows.

I grabbed exe2csv.exe and extracted the tables from the raw file which was uploaded and it saw 176 rows in the `{DD6636C4-8929-4683-974E-22C046A43763}` table.  I then directly used the ese parser tool (https://github.com/Velocidex/go-ese) and dumped out the catalog (`eseparser catalog srudb.dat `) and then dumped the actual table (`eseparser dump srudb.dat '{DD6636C4-8929-4683-974E-22C046A43763}'`)

This produced the correct number of rows so the parser appears to be working fine to me. The artifact however returns no rows. But this is because it tries to filter it by a regex https://github.com/Velocidex/velociraptor/blob/2dffd815e5de3bdd6499e65fe347ffdd1f6d21cc/artifacts/definitions/Windows/Forensics/SRUM.yaml#L81

The filter looks for the App as resolved by the AppId from the SruDbIdMapTable but on my system all the entries have an AppId 1 and UserId 2 and these do not have a name in the SruDbIdMapTable. So I think the issue is that the default regex requires a single char match (it is `.`) but the App resolves to an empty string. Setting ExecutableRegex to `.*` allows all rows to be displayed - on my system it seems that this field is just not set to anything sensible for this table (the other tables have normal things).

**Comment by chris-counteractive** [comment]
Brilliant - glad to see one mystery solved! A one-character mod to the default regex is the kind of fix you like to see.

As requested:

1. `ntfs.exe runs` for 59957: [ntfs-runs-59957.txt](https://github.com/Velocidex/velociraptor/files/5435258/ntfs-runs-59957.txt)
1. `ntfs.exe stat` for 59957: [ntfs-stat-59957.txt](https://github.com/Velocidex/velociraptor/files/5435310/ntfs-stat-59957.txt)
1. `ntfs.exe stat` for 19153: [ntfs-stat-19153.txt](https://github.com/Velocidex/velociraptor/files/5435314/ntfs-stat-19153.txt)
1. `ntfs.exe stat` for 59934: [ntfs-stat-59934.txt](https://github.com/Velocidex/velociraptor/files/5435315/ntfs-stat-59934.txt)
1. `ntfs.exe stat` for 329708: [ntfs-stat-329708.txt](https://github.com/Velocidex/velociraptor/files/5435316/ntfs-stat-329708.txt)
1. `ntfs.exe stat` for 24121: [ntfs-stat-24121.txt](https://github.com/Velocidex/velociraptor/files/5435317/ntfs-stat-24121.txt)

Here is how the two different accessors on this file compare using the VFS browser:
- `file`:  ![image](https://user-images.githubusercontent.com/48190665/97113866-07c55380-16bb-11eb-99a2-723beb00fac3.png)
- `ntfs`: ![image](https://user-images.githubusercontent.com/48190665/97113856-efedcf80-16ba-11eb-9a49-8a7520a324c7.png)

Some notes that might offer some clues or expose other issues:

1. The `ntfs` VFS tree shows `c:/Windows/System32/sru/SRUDB.dat` as mft entry 59957-128-**0**.  On the other hand, fls.exe shows it as  59957-128-**6**.
1. The M and A times are identical, C on NTFS is the same as M and A but differs from the `file` accessor C.
1. In v0.5.1 I can't download the file at all via the download link (in either tree) - it just looks like a no-op, no errors in the console.

Hope this helps, happy to keep working through the debug journey.  I never asked, but were you able to repro the initial issue or is it just me? 😃 Thanks again!

**Comment by scudette** [comment]
Hi Chris - thanks for the debugging work! This is perfect. You can see that there is a problem with joining up the VCNs in the ntfs parser.  The issue is really here:
```
Run 1 (*parser.MappedReader):
  Mapping 9060352 -> 9060352 with *parser.RangeReader
    RangeReader with 1 runs:
    Run 0 (*parser.MappedReader):
      Mapping 0 -> 6602752 with *parser.PagedReader
```

This run corresponds to the VCN but it is of length 0! So it gets messed up.

I submitted a fix and I also added a recorder feature to the ntfs.exe tool - you can get the latest version from https://github.com/Velocidex/go-ntfs/actions

The idea with the recorder is that we can replay only the relevant sectors from the disk image into the parser. This way we can collect various tests of interesting NTFS artifacts very cheaply (that is without storing a full disk image - most recordings are around 20-30kb). What happens is that it saves each cluster the parser touches in a cache directory and then anyone can replay the same thing - for example to record:

```
ntfs.exe stat \\.\c: 68310 --record dirname
```

Then on another system (even on linux) should produce the exact same output:
```
ntfs stat /dev/null 68310 --record dirname
```

It just parses the mft entry but as a side effect it saves all the clusters it touched. Then we can just store them in the repo and run tests against the same data - the parser will be able to get the recorded clusters from the repo and then we can test for regression and fix bugs etc. 

See https://github.com/Velocidex/go-ntfs/tree/master/tests/large_file_small_init for an example of how it works.

Anyway if you can check out the latest CI build and see if that works any better. If it still has issues I would appreciate if you also added the --record flag to the above commands and sent me the clusters then I can add it to the repo as a test as well.

**Comment by chris-counteractive** [comment]
Confirmed fixed for me as of 62314e8, both accessors produce the same results (and I noticed that GUI text-box reset bug is fixed, along with the updated default regex).  Hopefully this helps with other sparse collections (USN journal, etc.), it also dispels the confusion I was having about inconsistencies between the accessors.  I appreciate your help!

Glad to have contributed to the diagnosis, thanks for guiding me through it and turning things around so quickly.
